### ISSUE TITLE
CI: GH Actions encounter "No space left on device" unless unused CI tools are removed

### ISSUE THREAD

**Issue by connortann** [shap/shap#3330]
Our GitHub Actions workflow to run the unit tests has started failing. When trying to install pip-dependencies we now get an error `OSError: [Errno 28] No space left on device`:

```
Installing collected packages: wcwidth, sentencepiece, pytz, py4j, pure-eval, ptyprocess, pickleshare, mpmath, libclang, flatbuffers, executing, backcall, wrapt, wheel, urllib3, tzdata, typing-extensions, traitlets, tqdm, threadpoolctl, termcolor, tensorflow-io-gcs-filesystem, tensorflow-estimator, tensorboard-data-server, tenacity, sympy, slicer, six, setuptools, safetensors, regex, pyyaml, pyspark, pyparsing, pygments, pyasn1, protobuf, prompt-toolkit, pluggy, pillow, pexpect, parso, packaging, oauthlib, nvidia-nvtx-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufft-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, numpy, networkx, MarkupSafe, markdown, llvmlite, kiwisolver, keras, joblib, iniconfig, idna, grpcio, graphviz, gast, fsspec, fonttools, filelock, decorator, cycler, coverage, cloudpickle, charset-normalizer, certifi, cachetools, absl-py, werkzeug, triton, scipy, rsa, requests, python-dateutil, pytest, pyasn1-modules, plotly, opt-einsum, opencv-python, nvidia-cusparse-cu12, nvidia-cudnn-cu12, numba, ml-dtypes, matplotlib-inline, Jinja2, jedi, h5py, google-pasta, contourpy, astunparse, asttokens, xgboost, stack-data, scikit-learn, requests-oauthlib, pytest-cov, pandas, nvidia-cusolver-cu12, matplotlib, lightgbm, huggingface-hub, google-auth, torch, tokenizers, shap, pytest-mpl, pyod, ipython, gpboost, google-auth-oauthlib, catboost, transformers, torchvision, tensorboard, tensorflow
  Attempting uninstall: setuptools
    Found existing installation: setuptools 65.5.0
    Uninstalling setuptools-65.5.0:
      Successfully uninstalled setuptools-65.5.0
ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device
```

As an example, see tests runs on #3322 . Here's an [example job](https://github.com/shap/shap/actions/runs/6482905945/job/17672222943?pr=3322).

I haven't yet determined the root cause of the issue.

FYI @thatlittleboy , @CloseChoice , @dsgibbons it looks like this issue will block new PRs.

**Comment by connortann** [comment]
This could be useful perhaps
https://github.com/marketplace/actions/free-disk-space-ubuntu

**Comment by CloseChoice** [comment]
Should we investigate this further or even add a CI task that logs the disk of the virtualenv? There might be two reasons I can think of that caused this error:
 - some of our dependencies increased in size
 - ubuntu-latest comes with more tools installed, hence less space

**Comment by thatlittleboy** [comment]
It would be interesting to know the exact reason why this is happening suddenly now, but I personally wouldn't sink too much time into this.

I'm slightly more concerned about whether the items that we're deleting from the runner with the new action will affect current or  future CI runs, or whether it will result in perplexing bugs in our CI in the future.

**Comment by connortann** [comment]
The failing tests were fixed by #3331 - thanks @CloseChoice !

It would be useful to identify the root cause of the space issue. I wonder if something strange is going on: perhaps here is an issue with our dependencies. The current fix adds 2 minutes to our CI time, which is fine but isn't ideal.

That said, it's not a huge priority at the moment, the main thing is that our tests pass reliably.
